ClassSelectors for interfaces should be ignored

When a `ClassSelector` for an interface type is passed as part of a discovery request, discovery currently fails with an exception:

```
Caused by: org.junit.platform.commons.JUnitException: TestEngine with ID 'testng' failed to discover tests
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:111)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:85)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:92)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	... 25 more
Caused by: org.junit.platform.commons.JUnitException: ClassSelector [className = 'com.netflix.nebula.helloworld.HelloReplyOrBuilder'] resolution failed
	at org.junit.platform.launcher.listeners.discovery.AbortOnFailureLauncherDiscoveryListener.selectorProcessed(AbortOnFailureLauncherDiscoveryListener.java:39)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolveCompletely(EngineDiscoveryRequestResolution.java:102)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.run(EngineDiscoveryRequestResolution.java:82)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolver.resolve(EngineDiscoveryRequestResolver.java:113)
	at org.junit.support.testng.engine.TestNGTestEngine.discover(TestNGTestEngine.java:78)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:103)
	... 32 more
Caused by: java.lang.NullPointerException
	at org.junit.support.testng.engine.TestAnnotationUtils.getClassHierarchy(TestAnnotationUtils.java:84)
	at org.junit.support.testng.engine.TestAnnotationUtils.collectTestAnnotations(TestAnnotationUtils.java:77)
	at org.junit.support.testng.engine.TestAnnotationUtils.collectGroups(TestAnnotationUtils.java:72)
	at org.junit.support.testng.engine.TestDescriptorFactory.createClassDescriptor(TestDescriptorFactory.java:38)
	at org.junit.support.testng.engine.TestNGSelectorResolver.lambda$resolve$0(TestNGSelectorResolver.java:45)
```

Instead, interfaces and abstract classes should be ignored.

--- a/src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java
+++ b/src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java
@@ -10,13 +10,16 @@
 
 package org.junit.support.testng.engine;
 
+import static java.util.Spliterators.spliteratorUnknownSize;
+
 import java.lang.reflect.Method;
-import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.List;
+import java.util.Iterator;
 import java.util.Objects;
 import java.util.Optional;
+import java.util.Spliterator;
 import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
 
 import org.testng.ITestNGMethod;
 import org.testng.annotations.Test;
@@ -74,16 +77,27 @@ class TestAnnotationUtils {
 	}
 
 	private static Stream<Test> collectTestAnnotations(Class<?> testClass) {
-		return getClassHierarchy(testClass).stream() //
+		return getClassHierarchy(testClass) //
 				.map(clazz -> clazz.getAnnotation(Test.class)) //
 				.filter(Objects::nonNull);
 	}
 
-	private static List<Class<?>> getClassHierarchy(Class<?> testClass) {
-		List<Class<?>> result = new ArrayList<>();
-		for (Class<?> clazz = testClass; clazz != Object.class && clazz != null; clazz = clazz.getSuperclass()) {
-			result.add(clazz);
-		}
-		return result;
+	private static Stream<Class<?>> getClassHierarchy(Class<?> testClass) {
+		Iterator<Class<?>> iterator = new Iterator<Class<?>>() {
+			Class<?> next = testClass;
+
+			@Override
+			public boolean hasNext() {
+				return next != Object.class && next != null;
+			}
+
+			@Override
+			public Class<?> next() {
+				Class<?> result = next;
+				next = next.getSuperclass();
+				return result;
+			}
+		};
+		return StreamSupport.stream(spliteratorUnknownSize(iterator, Spliterator.ORDERED), false);
 	}
 }

--- a/src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java
+++ b/src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java
@@ -81,7 +81,7 @@ class TestAnnotationUtils {
 
 	private static List<Class<?>> getClassHierarchy(Class<?> testClass) {
 		List<Class<?>> result = new ArrayList<>();
-		for (Class<?> clazz = testClass; clazz != Object.class; clazz = clazz.getSuperclass()) {
+		for (Class<?> clazz = testClass; clazz != Object.class && clazz != null; clazz = clazz.getSuperclass()) {
 			result.add(clazz);
 		}
 		return result;

--- a/src/test/java/org/junit/support/testng/engine/DiscoveryIntegrationTests.java
+++ b/src/test/java/org/junit/support/testng/engine/DiscoveryIntegrationTests.java
@@ -263,7 +263,8 @@ class DiscoveryIntegrationTests extends AbstractIntegrationTests {
 	}
 
 	@ParameterizedTest
-	@ValueSource(classes = { InterfaceTestCase.class, AbstractTestCase.class })
+	@ValueSource(classes = { InterfaceTestCase.class, AbstractTestCase.class, RecordTestCase.class,
+			EnumTestCase.class })
 	void doesNotThrowExceptionWhenNonExecutableTypeOfClassIsSelected(Class<?> testClass) {
 		var request = request().selectors(selectClass(testClass)).build();
 
@@ -276,5 +277,11 @@ class DiscoveryIntegrationTests extends AbstractIntegrationTests {
 	}
 
 	static abstract class AbstractTestCase {
+	}
+
+	record RecordTestCase() {
+	}
+
+	enum EnumTestCase {
 	}
 }

--- a/src/test/java/org/junit/support/testng/engine/DiscoveryIntegrationTests.java
+++ b/src/test/java/org/junit/support/testng/engine/DiscoveryIntegrationTests.java
@@ -261,4 +261,20 @@ class DiscoveryIntegrationTests extends AbstractIntegrationTests {
 		TestDescriptor methodDescriptor = getOnlyElement(classDescriptor.getChildren());
 		assertThat(methodDescriptor.getDisplayName()).isEqualTo("test");
 	}
+
+	@ParameterizedTest
+	@ValueSource(classes = { InterfaceTestCase.class, AbstractTestCase.class })
+	void doesNotThrowExceptionWhenNonExecutableTypeOfClassIsSelected(Class<?> testClass) {
+		var request = request().selectors(selectClass(testClass)).build();
+
+		var rootDescriptor = testEngine.discover(request, engineId);
+
+		assertThat(rootDescriptor.getChildren()).isEmpty();
+	}
+
+	interface InterfaceTestCase {
+	}
+
+	static abstract class AbstractTestCase {
+	}
 }

